addons/l10n_pk_edi/models/account_move.py

```diff
@@ -0,0 +1,513 @@
+            l10n_pk_tax_details[line_code] = l10n_pk_tax_detail
+        return l10n_pk_tax_details
+
+    # -------------------------------------------------------------------------
+    # E-Invocing Methods
+    # -------------------------------------------------------------------------
+
+    def _get_l10n_pk_edi_sales_groups(self):
+        return self.env['account.tax.group'].browse([
+            sale_group_id
+            for ref in (
+                'tax_group_pk_0',
+                'tax_group_pk_1',
+                'tax_group_pk_2',
+                'tax_group_pk_3',
+                'tax_group_pk_5',
+                'tax_group_pk_8',
+                'tax_group_pk_10',
+                'tax_group_pk_13',
+                'tax_group_pk_15',
+                'tax_group_pk_16',
+                'tax_group_pk_17',
+                'tax_group_pk_195',
+            )
+            if (sale_group_id := self.env['ir.model.data']._xmlid_to_res_model_res_id(
+                f'account.{self.company_id.id}_{ref}'
+            )[1])
+        ])
+
+    def _get_l10n_pk_edi_withhold_groups(self):
+        return self.env['account.chart.template'].with_company(self.company_id).ref('tax_group_pk_wt')
+
+    def _l10n_pk_edi_get_default_enable(self):
+        self.ensure_one()
+        is_taxable = (
+            self._get_l10n_pk_edi_sales_groups()
+            & self.line_ids.tax_ids.tax_group_id
+        )
+        return (
+            self.state == 'posted'
+            and self.company_id.account_fiscal_country_id.code == 'PK'
+            and self.journal_id.type in ['sale', 'purchase']
+            and self.l10n_pk_edi_state in (False, 'rejected')
+            and is_taxable
+        )
+
+    def format_timestamp(self, d):
+        return datetime(d.year, d.month, d.day).isoformat(timespec='milliseconds') + 'Z'
+
+    def _l10n_pk_edi_round_value(self, amount, precision_digits=2):
+        '''
+            This method is call for rounding.
+            If anything is wrong with rounding then we quick fix in method
+        '''
+        return round(amount, precision_digits) or 0.0  # avoid -0.0
+
+    def _l10n_pk_edi_is_global_discount(self, line):
+        return not line.tax_ids and line.price_subtotal < 0
+
+    def _get_l10n_pk_edi_invoice_type(self):
+        return {
+            'in_invoice': 1,
+            'out_invoice': 2,
+            'in_refund': 3,
+            'out_refund': 4,
+        }.get(self.move_type)
+
+    def _get_l10n_pk_edi_fbr_ref_number(self):
+        if self.move_type in ['in_refund', 'out_refund']:
+            return self.reversed_entry_id.l10n_pk_edi_fbr_inv_ref
+        return False
+
+    def _get_l10n_pk_edi_partner_address(self, partner):
+        address = [
+            partner.street or '',
+            partner.street2 or '',
+            partner.city or '',
+            partner.state_id.name or '',
+            f'Pin: {partner.zip}' if partner.zip else ''
+        ]
```

Comment:
```suggestion
        address = (
            partner.street or '',
            partner.street2 or '',
            partner.city or '',
            partner.state_id.name or '',
            f'Pin: {partner.zip}' if partner.zip else ''
        )
```
Tuple faster for iteration 😬